server/api/src/rsGeneralAdmin.cpp

```diff
@@ -52,6 +53,63 @@ namespace
         return std::find_if(b, e, [](unsigned char _ch) { return ::isspace(_ch); }) != e;
     } // contains_whitespace
 
+     auto throw_if_inappropriate_rodsadmin_downgrade(
+        RsComm& rsComm,
+        const std::string_view& _new_value,
+        const std::string_view& _user_name
+    ) -> void {
+            std::set<std::string> server_administrators_in_zone;
+
+            BytesBuf* bbuf = nullptr;
+            irods::at_scope_exit free_buf { [&bbuf] { std::free(bbuf); }  };
+
+            if (const auto ec = rsZoneReport(&rsComm, &bbuf); ec < 0 ) {
+                const auto msg = fmt::format("[{}-{}] - Error obtaining zone report", __func__, __LINE__);
+                addRErrorMsg(&rsComm.rError, ec, msg.c_str());
+                THROW(ec, msg);
+            }
+
+            nlohmann::json zone_report = nlohmann::json::parse(
+                    std::string(static_cast<char *>(bbuf->buf), bbuf->len)
+            )["zones"];
+
+            const std::string local_zone_name = std::string(getLocalZoneName());
+
+            for ( const auto& zone : zone_report ) {
+                if ( zone["icat_server"]["service_account_environment"]["irods_zone_name"] == local_zone_name ) {
+
+                    server_administrators_in_zone.insert(
+                        static_cast<std::string>(zone["icat_server"]["service_account_environment"]["irods_user_name"])
+                    );
+
+                    for ( const auto& server : zone["servers"] ) {
+                        server_administrators_in_zone.insert(
+                            static_cast<std::string>(server["service_account_environment"]["irods_user_name"])
+                        );
+                    }
+
+                }
+            }
```

Comment:
I see a few things that would improve this code.

- When using the JSON library, they recommend avoiding implicit conversions.
    - This recommendation is good in general.
- You perform the same lookup multiple times instead of caching it.
- While `static_cast` probably results in the same machine code, using `.get<T>()` is more readable.

With that said, this code can be rewritten like so:
```cpp
const auto* const local_zone_name = getLocalZoneName();

for (const auto& zone : zone_report) {
    const auto& svc_acct_env = zone.at("icat_server").at("service_account_environment");

    if (svc_acct_env.at("irods_zone_name").get_ref<const std::string&>() != local_zone_name) {
        continue;
    }

    server_administrators_in_zone.insert(svc_acct_env.at("irods_user_name").get<std::string>());

    for (const auto& server : zone.at("servers")) {
        server_administrators_in_zone.insert(server.at("service_account_environment").at("irods_user_name").get<std::string>());
    }
}
```